UI: Add ability to reorder mixer by drag & drop

cg2121 opened the pull request:
### Description
This adds the ability for users to drag and drop items in the audio mixer to reorder them. This changes the mixer to a QListWidget.

### Motivation and Context
https://ideas.obsproject.com/posts/117/add-the-ability-to-reorder-audio-mixer-items

### How Has This Been Tested?
Moving around mixer items.

### Types of changes
- New feature (non-breaking change which adds functionality)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.

WizardCM commented:
Alright, initial test.

1. Not a fan of the light grey 2px solid border, feels very out of place. I'd go with the styling we have in the Scenes list if possible
2. Currently dragging only really works if you drag the top row (name/gap/dB), which isn't particularly obvious. See if it's possible to make the volume meter & ticks draggable too (as I expect that's where people will try to grab most). If you can go one step further and do the same for the  volume/options buttons, that'd be nice to have, but I find that really unlikely (and less used).
3. Other than when dragging, I don't think they should be "selectable", unless you plan on making Ctrl+Up/Down reorder (which, actually thinking about it, we should - control via keyboard is great for accessibility)
4. While dragging, the debug output log seems to get spammed with `onecore\com\combase\dcomrem\resolver.cxx(2283)\combase.dll!00007FFD8F0AE3BE: (caller: 00007FFD8F113206) ReturnHr(5849) tid(ed8c) 80040154 Class not registered` - not quite sure why.

Overall, super excited to see this.

cg2121 commented:
1. I chose the border because this is what it would look like otherwise:
![Screenshot from 2020-04-01 23-45-11](https://user-images.githubusercontent.com/19962531/78211556-eb3da180-7472-11ea-9f1d-1973e2ae1093.png)

2. As far as I can tell this wouldn't be possible.

WizardCM commented:
Heads up: with the latest force-push, the default Desktop Audio and Mic/Aux don't appear.

cg2121 commented:
@WizardCM the issue you mentioned about Desktop and Mic/Aux sources not showing up, is fixed now.

WizardCM commented:
@cg2121 Actually, behaves even weirder.

I had Desktop Audio and Mic/Aux set to Default. Mic/Aux showed up fine, but Desktop Audio didn't. So I went in and manually set both. which made both appear.. as duplicates.

![image](https://user-images.githubusercontent.com/941350/78973013-5877c380-7b52-11ea-8e16-0d782e7b0b20.png)

WizardCM commented:
> > See if it's possible to make the volume meter & ticks draggable too (as I expect that's where people will try to grab most).

> As far as I can tell this wouldn't be possible.

I figured it out!

Add
```cpp
volControl->volMeter->setEnabled(false);
```
After this line:

https://github.com/obsproject/obs-studio/blob/master/UI/volume-control.cpp#L34

Which makes the volMeter ignore mouse events, allowing for much easier drag/drop.

cg2121 commented:
@WizardCM could you check now if the newest force-push fixes your issues?

I've also implemented your idea of disabling the volume meter, so users can click on it to drag.

WizardCM commented:
Can confirm, with the latest changes, this is quite stable.

I still have a few UX concerns:
* Once you've selected an item in the mixer, there is no way to deselect it. One entry is now always permanently a different colour. This is not the case for the Scenes or Sources list - the moment you focus another element, they get unfocused. I definitely still think that there should only be a visual indicator while the item is being dragged.
* While I see the logic behind making the source order per-scene, I also worry about users who have lots of scenes. Having to go through every single scene just to move the mic source to the top seems like a waste of time, and then items shifting unexpectedly when toggling between scenes is not a good experience. I vote the order be global.


Finally, as mentioned in my original comment, the debug log gets spammed while the mixer item is being dragged. This does not occur with anything else, so I can only assume it's fixable.

cg2121 commented:
> I vote the order be global.

What if the user has a audio source that is in only one scene? How would it work if they moved it to the top?

> Having to go through every single scene just to move the mic source to the top 

Initially for each scene, the global sources would all be in the same order.

> the debug log gets spammed

This doesn't happen for me when testing on Windows.

cg2121 commented:
I've now changed it to where clicking anywhere on the window clears the highlighting. I've also changed it to when adding new audio sources, they are added to the top instead of the bottom.

unknowndomain commented:
In talking to someone in #ideas-suggestions this evening it came to my mind that some people (myself included) might have liked the original behaviour (Alphabetical), if this hasn't already been implemented can we have a button in the right click contextual menu to sort the mixer alphabetically.

Warchamp7 commented:
Few notes!

I agree with Matt this sorting should be a global order, not per scene.

I also think the hover styling should only show up when hovering an area that can actually be dragged to reorder
Ex. Hovering over the volume meter or mute icon should not show the hover.

I'm currently unsure how best to solve this since I haven't looked at how a mixer item is constructed widget-wise, but it may be a matter of targeting the QLabel used for title and decibel total. Might need a frame around them? Not sure.

![image](https://user-images.githubusercontent.com/1554753/131405897-9ed1d001-6c0b-4ad1-87ef-64cace7411be.png)
_Mockup_

I don't think the meters themselves should be draggable to create a clear gap between targetable areas. 

![image](https://user-images.githubusercontent.com/1554753/131406036-4ee4a84a-ffa5-408e-90db-619d59885d53.png)
_If I click here and drag am I going to reorder this mixer item or am I going to change it's volume?_

Two bugs I've noticed:
https://user-images.githubusercontent.com/1554753/131405289-57b2a25d-ea47-4aae-a624-842b8a3d9767.mp4

- Dragging a mixer item sometimes causes it to become invisible
- Order does not seem to be retained when switching off the scene and back

cg2121 commented:
Updated to latest master.

cg2121 commented:
This needs a custom list model to work properly, similar to the source tree, so I am closing for now.
